REST-API-Basics/web/src/main/java/com/epam/esm/web/controller/GiftCertificateController.java

```diff
@@ -27,64 +31,38 @@ public GiftCertificateController(GiftCertificateService giftCertificateService)
 
     @GetMapping
     @ResponseStatus(HttpStatus.OK)
-    public List<GiftCertificate> getAll() {
-        return giftCertificateService.getAll();
-    }
-
-    @GetMapping("/with_tags")
-    @ResponseStatus(HttpStatus.OK)
-    public List<GiftCertificateDto> getAllWithTags() {
-        return giftCertificateService.getAllWithTags();
+    public List<GiftCertificate> getAll(
+            @RequestParam(name = "giftCertificateId", required = false) Long giftCertificateId
+    ) {
+        return giftCertificateService.getRoute(giftCertificateId);
     }
 
-    @GetMapping("/{giftCertificateId}")
-    @ResponseStatus(HttpStatus.OK)
-    public GiftCertificate getById(@PathVariable("giftCertificateId") Long id) throws ServiceException {
-        return giftCertificateService.findById(id);
+    @GetMapping("/with-tags")
+    public List<GiftCertificateDto> getAllWithTags(
+            @RequestParam(name = "tagName", required = false) String tagName,
+            @RequestParam(name = "sort", required = false) List<String> sortColumns,
+            @RequestParam(name = "order", required = false) List<String> orderType,
+            @RequestParam(name = "filter", required = false) List<String> filterBy
+    ) {
+        return giftCertificateService.getRouteWithTags(tagName, sortColumns, orderType, filterBy);
     }
 
-    @PutMapping("/{giftCertificateId}")
+    @PutMapping
     @ResponseStatus(HttpStatus.OK)
-    public GiftCertificateDto updateById(@PathVariable("giftCertificateId") Long id,
-                                         @RequestBody GiftCertificateDto giftCertificateDto) throws ServiceException {
+    public GiftCertificateDto updateById(@RequestParam("giftCertificateId") Long id,
```

Comment:
id is a selector for resource, use it in the path if you operating with specific certificate. Here and in the other endpoints